appintro/src/main/java/com/github/appintro/internal/AppIntroViewPagerController.kt

```diff
@@ -0,0 +1,283 @@
+import kotlin.math.max
+
+/**
+ * Class that controls the [ViewPager2] of AppIntro.
+ * This is responsible of handling of paging, managing touch and dispatching events.
+ *
+ * @property isFullPagingEnabled Enable or disable swiping at all.
+ * @property isPermissionSlide If the current slide has permissions.
+ * @property onNextPageRequestedListener Listener for Next Page events.
+ */
+internal class AppIntroViewPagerController(
+    private val viewPager: ViewPager2,
+    private val viewPagerGestureOverlay: GestureOverlayView
+) {
+
+    var isFullPagingEnabled = true
+    var isPermissionSlide = false
+    var onNextPageRequestedListener: AppIntroViewPagerListener? = null
+
+    private var currentTouchDownX: Float = 0.toFloat()
+    private var currentTouchDownY: Float = 0.toFloat()
+    private var illegallyRequestedNextPageLastCalled: Long = 0
+    private var pageChangeCallback: AppIntroBase.OnPageChangeCallback? = null
+
+    init {
+        addPagerTouchInterceptor()
+    }
+
+    fun goToNextSlide() {
+        with(viewPager) {
+            // avoid IllegalStateException when changing item while fake dragging
+            endFakeDrag()
+            setCurrentViewPagerItem(
+                position = if (!LayoutUtil.isRtl(context)) currentItem + 1 else currentItem - 1,
+                smoothScrool = true
+            )
+        }
+    }
+
+    fun goToPreviousSlide() {
+        with(viewPager) {
+            // avoid IllegalStateException when changing item while fake dragging
+            endFakeDrag()
+            setCurrentViewPagerItem(
+                position = if (!LayoutUtil.isRtl(context)) currentItem - 1 else currentItem + 1,
+                smoothScrool = true
+            )
+        }
+    }
+
+    fun isFirstSlide(size: Int): Boolean {
+        with(viewPager) {
+            return if (LayoutUtil.isRtl(context)) (currentItem - size + 1 == 0) else (currentItem == 0)
+        }
+    }
+
+    fun isLastSlide(size: Int): Boolean {
+        with(viewPager) {
+            return if (LayoutUtil.isRtl(context)) (currentItem == 0) else (currentItem - size + 1 == 0)
+        }
+    }
+
+    fun getCurrentSlideNumber(size: Int): Int {
+        with(viewPager) {
+            return if (LayoutUtil.isRtl(context)) (size - currentItem) else (currentItem + 1)
+        }
+    }
+
+    /**
+     * Override is required to trigger [AppIntroBase.OnPageChangeCallback.onPageSelected] for the first page.
+     * This is needed to correctly handle progress button display after rotation on a locked first page.
+     */
+    fun setCurrentViewPagerItem(position: Int, smoothScrool: Boolean = false) {
+        with(viewPager) {
+            endFakeDrag()
+
+            val oldItem = currentItem
+            viewPager.setCurrentItem(position, smoothScrool)
+
+            // When you pass set current item to 0,
+            // The pageChangeListener won't be called so we call it on our own
```

Comment:
Why this?

Reply:
This should be needed to trigger the pageChangeListener callback on the first slide (since the page isn't really changed, but it's the first one loaded).

That's just in case the first page has a custom policy logic that needs to run on that callback.